sdk/webpubsub-socketio-extension/src/SIO/components/negotiate.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  debugModule,
+  AzureSocketIOOptions,
+  AzureSocketIOCredentialOptions,
+  getWebPubSubServiceClient,
+  ConfigureNegotiateOptions,
+  NegotiateResponse,
+  writeResponse
+} from "../../common/utils";
+import { WEB_PUBSUB_OPTIONS_PROPERY_NAME } from "./constants";
+import * as SIO from "socket.io";
+import { Store, Session } from "express-session";
+import { Request as ExpressRequest, Response as ExpressResponse } from "express";
+import { parse } from "url";
+
+const debug = debugModule("wps-sio-ext:SIO:negotiate");
+
+/**
+ * Returns a Express middleware to handle negotiate request whose url starts with `path`
+ *
+ * @param path - the path of negotiate request. For Example `/negotiate`.
+ * @param io - the Socket.IO server or its pure options
+ * @param configureNegotiateOptions - a customized function which defines how to extract information for negotiation from a HTTP request
+ * @returns
+ */
+export function negotiate(
+  io: SIO.Server | AzureSocketIOOptions | AzureSocketIOCredentialOptions,
+  configureNegotiateOptions: ConfigureNegotiateOptions
+): (req: ExpressRequest, res: ExpressResponse, next: any) => void {
+  debug(`getNegotiateExpressMiddleware`);
+
+  const wpsOptions: AzureSocketIOOptions | AzureSocketIOCredentialOptions =
+    io instanceof SIO.Server ? io[WEB_PUBSUB_OPTIONS_PROPERY_NAME] : io;
+  const serviceClient = getWebPubSubServiceClient(wpsOptions);
+
+  return async (req: ExpressRequest, res: ExpressResponse, next: any): Promise<void> => {
```

Comment:
Can we just use `IncomingMessage` here as we don't use any special properties in `express.Request`

Reply:
updated. `express.Request` inherits `IncomingMessage`